phylum-dev/phylum-ci
# feat!: allow `.phylum_project` file to be optional

Actions taken include:

* Add support for multiple lockfiles
  * Bump the minimum required Phylum CLI version to v4.5.0
    * Ability to specify multiple lockfiles was added then
  * Create a `Lockfile` class to represent a single lockfile
    * Move properties of a lockfile from `CIBase` to `Lockfile` class
  * Update each CI environment to update every lockfile's change status
* Consolidate initialization/instantiation logic in class properties
  * The Phylum project name
    * Generate a deterministic project name when not otherwise provided
  * The Phylum group name
  * The lockfile(s) in use
    * Detect the lockfile(s) when not otherwise provided
  * The Phylum CLI path
  * Lockfile hash object
* Move "ensure a project" function from cli to a `CIBase` class method
* Remove pre-requisite check for `.phylum_project` file
* Add restrictions to the Phylum label generation
  * Start the label with the `self.ci_platform_name`
  * Replace all runs of whitespace characters with single `-` character
  * Update each CI implementation to adhere to the restrictions
* Add  `git_repo_name` function to get a git repository's name
  * Refactor git helper functions to allow for easier testing
  * Add tests for the new function
* Add `pathspec` to main dependencies
  * Used to respect ignore files when detecting lockfiles
* Add `dulwich` to test dependencies
  * Used to test `git` functionality programmatically w/o `subprocess`
* Bump the PyPI package development status from `Alpha` to `Beta`
* Update documentation
  * Allow for the possibility of multiple lockfiles
  * Add examples for how to specify multiple lockfiles
  * Remove wording specifing the `.phylum_project` file as required
  * Add a reference to the release process in `CONTRIBUTING.md`
  * Update method for passing additional options through `poetry`/`tox`
  * Clean up formatting in `CHANGELOG.md`
* Format and refactor throughout
  * Check for 100 return code when getting `phylum analyze` results
  * Use source branch consistently in PR/MR labels
  * Use correct PR number in ADO labels for GitHub repos

Closes #147

BREAKING CHANGE: CLI installs prior to v4.5.0 are no longer supported.
A Phylum CLI version with ability to specify multiple lockfiles is
required.

## Checklist

- [x] Does this PR have an associated issue (i.e., `closes #<issueNum>` in description above)?
- [x] Have you ensured that you have met the expected acceptance criteria?
- [x] Have you created sufficient tests?
  - A handful of new unit tests were added to cover the new `git` helper function
  - Manual testing has been completed on the local (CINone) and pre-commit environments
  - Manual testing has been completed on the CI environments for: GitHub, Bitbucket, GitLab, and Azure Pipelines
  - The changes in this PR are available for testing as a Docker image: `docker pull maxrake/phylum-ci:project_shun`
- [x] Have you updated all affected documentation?
  - [x] Documentation in the `phylum-ci` repo has been updated
  - [x] Documentation in the `phylum-analyze-pr-action` repo has been updated
    - This should happen _after_ the changes here have been released
    - See https://github.com/phylum-dev/phylum-analyze-pr-action/pull/24

## Screenshots

---

The `pre-commit` environment was exercised by updating the lockfile with a known bad dep. In this case, a `.phylum_project` file was present to obtain the project name, group, and lockfile(s) in use.

<img width="1115" alt="image" src="https://user-images.githubusercontent.com/18729796/228053232-b866e9b3-32b7-4c16-99b8-c898b43ee0da.png">

---

The `CINone` environment was exercised by updating the lockfile with a known bad dep. In this case, a `.phylum_project` file was present to obtain the project name, group, and lockfile(s) in use.

<img width="1115" alt="image" src="https://user-images.githubusercontent.com/18729796/228054335-c13008ad-c6cc-4796-b891-3c2e7183a470.png">

---

Multiple lockfiles specified as options. One of them is an empty file and get correctly filtered out.

<img width="1111" alt="image" src="https://user-images.githubusercontent.com/18729796/228055175-a1541295-43fb-439b-8a37-d0dd29c45527.png">

---

Multiple lockfiles specified in the `.phylum_project` file. One is an empty file, another is non-existent, and a third is not actually a lockfile. These were all correctly filtered out.

<img width="1111" alt="image" src="https://user-images.githubusercontent.com/18729796/228056327-a44bc2cd-d3f0-4644-9164-750103b9d14b.png">

---

Multiple lockfiles, at different directory levels, and no `.phylum_project` file. The `.ignore` file is not a real ignore file (it is a binary that was renamed).

<img width="1111" alt="image" src="https://user-images.githubusercontent.com/18729796/228057771-61ceb1ba-fdf3-4482-943c-7e1bba253d9f.png">

---

Multiple lockfiles, at different directory levels, and no `.phylum_project` file. The `--force-analysis` flag is specified without the `--all-deps` flag.

<img width="1111" alt="image" src="https://user-images.githubusercontent.com/18729796/228068532-1b34d521-3ea7-434e-b845-9f80f52e5b86.png">

---

Logs from GitHub Action run during a PR. In this case, a `requirements.txt` file that already existed in the root of the repo was updated to add a known bad dep. Additionally, a new lockfile, also named `requirements.txt`, was added with some known bad deps...but 7 levels deep in the directory tree. The findings were as expected and the PR was correctly commented.

<img width="1146" alt="image" src="https://user-images.githubusercontent.com/18729796/228050760-9bfb1b34-28c9-4c83-b110-e89a5c04e7f4.png">


---

Logs from Bitbucket run in a branch pipeline.

<img width="2492" alt="image" src="https://user-images.githubusercontent.com/18729796/228073106-d06dee14-ad11-43a3-9329-a8f78667e34c.png">

---

Logs from GitLab run in a branch pipeline. The lockfile was specified on the command line since it has a non-standard name. The `.phylum_project` file was removed and the dependencies were updated to be current.

<img width="1296" alt="image" src="https://user-images.githubusercontent.com/18729796/228304266-aee88526-91f0-4f0d-9301-44e0052e433a.png">

---

Logs from GitLab run in a merge request pipeline. The lockfile was updated to add a few known bad deps. The comment was applied to the MR correctly.

<img width="1296" alt="image" src="https://user-images.githubusercontent.com/18729796/228308224-366aebce-d318-4b34-819d-d477b92a24a1.png">

---

Logs from Azure Pipelines run in a CI trigger pipeline context. The `.phylum_project` file was removed and the dependencies were updated to be current.

<img width="1794" alt="image" src="https://user-images.githubusercontent.com/18729796/228320892-517b39bd-2fb2-4020-b5ea-2522fe0ab38c.png">

---

Logs from Azure Pipelines run in XYZ pipeline context. The lockfile was updated to add a few known bad deps. The comment was applied to the PR correctly.

<img width="1794" alt="image" src="https://user-images.githubusercontent.com/18729796/228328175-7bd09c3b-3abf-4de3-b883-8c40cdd288db.png">

---
